compiler/rustc_trait_selection/src/solve/fulfill.rs

```diff
@@ -380,7 +380,12 @@ impl<'tcx> ProofTreeVisitor<'tcx> for BestObligation<'tcx> {
             source: CandidateSource::Impl(impl_def_id),
             result: _,
         } = candidate.kind()
-            && goal.infcx().tcx.has_attr(impl_def_id, sym::do_not_recommend)
+            && goal
+                .infcx()
+                .tcx
+                .get_attrs_by_path(impl_def_id, &[sym::diagnostic, sym::do_not_recommend])
+                .next()
+                .is_some()
```

Comment:
It might be worth to introduce a `.has_attrs()` method for this?

Reply:
Yeah, could you do that? `has_attrs_by_path`? (or `has_attrs_with_path`)?